### ISSUE TITLE
Single categorical column failure

### ISSUE THREAD

**Issue by david-waterworth** [catboost/catboost#617]
I've been experimenting with trivial problems to see how efficient different features are. I've been experimenting with approximating the function $sin(\omega t + \varphi)$

I've tried a few options to compare efficiency - i.e. two X columns sin(\omega t) and cos(\omega t), a single X column 0-2 $\pi$ etc. I figured as well it should work to have a single categorical column tabulated against the desired output. However the results are quite different between a single continuous feature (first case below, gives as expected a piece-wise constant approximation to $sin(\omega t + \varphi)$) and the second where t is a categorical variable with 1 level distinct level per sample.

In the second case it doesn't seem to grow even a stump and at each step the loss is unchanged? I understand this is a degenerate case but I'm interested in why. 

```
t = np.arange(0,24,1) 
w = 2*np.pi / 24 
p = -8*w
X = pd.DataFrame({"t" : t})
y = np.sin(w*t + p)
estimator = CatBoostRegressor(logging_level='Debug')
estimator.fit(X,y)
plt.plot(estimator.predict(X))
```
> Borders for float features generated
> Mem usage: Extract bools done: 165609472
> Mem usage: Before start train: 165609472
> 
> t, bin=4 score 2.379969516
> t, bin=13 score 3.064211416
> t, bin=13 score 3.299976742
>   tensor 1 is redundant, remove it and stop
> 0:	learn: 0.6948137	total: 3.34ms	remaining: 0us

```
X["t"].astype('category')
estimator = CatBoostRegressor(logging_level='Debug')
estimator.fit(X,y,cat_features=[0])
plt.plot(estimator.predict(X))
```

> age: Extract bools done: 166576128
> Mem usage: Before start train: 166576128
> 
> {t} pr0 tb0 type1, border=3 score -0.3552107792
>   tensor 0 is redundant, remove it and stop
> 0:	learn: 0.7071068	total: 2.68ms	remaining: 0us

**Comment by annaveronika** [comment]
The explanation is in the way we are treating categorical features. We generate new features based on categorical ones, which are representing estimates of probability of having positive value given that the object is of this category.

But we are not using this particular object when we are calculating the feature value for this object. We are using only the objects before this one in some random permutation. This helps a lot with fighting overfitting, because you are not memorizing the data. So if there is a single object with this category and nothing else, you cannot learn anything.

Here is a paper that explains this approach in more details:
http://learningsys.org/nips17/assets/papers/paper_11.pdf
You need Part 2 "Categorical features".

Hope it helps!
